compiler/rustc_hir/src/def.rs

```diff
@@ -47,33 +49,40 @@ pub enum NonMacroAttrKind {
     Registered,
 }
 
+/// What kind of definition something is; e.g., `mod` vs `struct`.
 #[derive(Clone, Copy, PartialEq, Eq, Encodable, Decodable, Hash, Debug)]
 #[derive(HashStable_Generic)]
 pub enum DefKind {
     // Type namespace
     Mod,
-    /// Refers to the struct itself, `DefKind::Ctor` refers to its constructor if it exists.
+    /// Refers to the struct itself, [`DefKind::Ctor`] refers to its constructor if it exists.
     Struct,
     Union,
     Enum,
-    /// Refers to the variant itself, `DefKind::Ctor` refers to its constructor if it exists.
+    /// Refers to the variant itself, [`DefKind::Ctor`] refers to its constructor if it exists.
     Variant,
     Trait,
-    /// `type Foo = Bar;`
+    /// Type alias: `type Foo = Bar;`
     TyAlias,
     ForeignTy,
+    /// Trait alias: `type IntIterator = Iterator<Item = i32>;`
     TraitAlias,
+    /// Associated type: `trait MyTrait { type Assoc; }`
     AssocTy,
+    /// Type parameter: the `T` in `Vec<T>`
     TyParam,
 
     // Value namespace
     Fn,
     Const,
+    /// Const-generic type parameter.
     ConstParam,
     Static,
     /// Refers to the struct or enum variant's constructor.
```

Comment:
It might help to give an example of when the Ctor is different from the type itself: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=af764b9af1c0b2373e9f8a28484231c2

Reply:
I don't think I understand the example well enough to document this; perhaps you or someone else could add some documentation in a follow-up? I would like to land this PR and then we can do follow-ups.

Reply:
Sure, seems good. I came up with a slightly more extended example in the meantime so I don't forget https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ed6253f734e68b2594bb6715886fa019 - tl;dr SelfCtor is in the value namespace but the type is in the type namespace.

Reply:
Oh, I can say something about how they're in different namespaces—is that what you meant?

Reply:
Yes, that's what I meant :)

Reply:
Hmm, now I feel confused: why do both `Variant` and `Ctor` exist? `Variant`s are in the value namespace, aren't they?

Reply:
Variants are in type namespace, https://github.com/rust-lang/rfcs/pull/2593 can give more intuition on why is this the case.
Variant constructors are values - callable (for tuple constructors) or ~not (for unit constructors).

Reply:
Ah, I was wondering if that RFC fit into this! I added an explanation based on what you described. 1e05d9e7753